Use "devtools" automation protocol as fallback

christian-bromann opened the pull request:
## Proposed changes

Given that Puppeteer now supports Chrome, Firefox and Edge (Chromium) without having to install a browser it makes sense to use it as fallback in case a driver doesn't exist. This patch adds:

- support for Edge (Chromium) using the devtools automation protocol
- removal of `puppeteer-firefox` dependency (now lives directly within `puppeteer`)
- driver endpoint check and fallback to `devtools` if endpoint couldn't be reached

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] I have added proper type definitions for new commands (if appropriate)

## Further comments

This is work in progress. It needs to be tested on Linux and Windows.

### Reviewers: @webdriverio/technical-committee

mgrybyk commented on packages/webdriverio/src/utils.js:
```diff
@@ -414,3 +415,33 @@ export const enhanceElementsArray = (elements, parent, selector, foundWith = '$$
+     * don't modify automation protocol if hostname is set as it is very likely
+     * that "webdriver" will be used
+     */
+    if (config.hostname || (config.user && config.key)) {
```
should we add config.port as well?

mgrybyk replied:
should we add config.port as well?

mgrybyk commented on packages/webdriverio/src/utils.js:
```diff
@@ -414,3 +415,33 @@ export const enhanceElementsArray = (elements, parent, selector, foundWith = '$$
+            method: 'GET',
+            host: 'localhost',
+            port: config.port || 4444,
+            path: '/status'
```
isn't config.path ignored in such case?

christian-bromann replied:
I will remove `config.port` - if nothing is specified we need to assume that the driver is on `localhost:4444`.

mgrybyk replied:
Yeah, but path /status is not necessarily valid. Can it be /wd/hub/path ?

christian-bromann replied:
`/status` is defined by the WebDriver spec and therefor has to be there

christian-bromann commented:
@webdriverio/project-committers 

This is finally ready for review, here is the Gist of things that change with this patch:

- I updated the Travis build to run lint, unit and smoke test separated from e2e (given that they require a different setup where browser need to be installed)

Build before:
![Screenshot 2020-02-15 at 18 37 54](https://user-images.githubusercontent.com/731337/74592511-66153f00-5022-11ea-86af-c9bff0c4e1c1.png)

Build now:
![Screenshot 2020-02-15 at 18 37 36](https://user-images.githubusercontent.com/731337/74592516-6c0b2000-5022-11ea-9c23-5f08477e96bd.png)

Given that we have 5 concurrent VMs the overall time shouldn't change much.

- I added support for Chromium Edge (currently only Mac and Windows)
- The DevTools automation is now default, that means if WebdriverIO can't find a driver endpoint it defaults to that
- due to a weird bug in Jest I had to change the setup for test coverage which now always includes all files (even the ones that are not test)

mgrybyk commented on .travis.yml:
```diff
@@ -1,18 +1,91 @@
-  - 10
-  - 12
+cache:
+  directories:
```
do we need to cache website's node modules here?

mgrybyk commented:
is it due to Jest' bug there are some index files ignored `/* istanbul ignore file */` ?

mgrybyk commented on packages/devtools/package.json:
```diff
@@ -35,12 +35,9 @@
+    "puppeteer-core": "^2.1.1"
   },
   "devDependencies": {
     "@types/puppeteer": "^1.19.0"
```
let's update types to 2.0.0 ?

mgrybyk commented on packages/webdriverio/src/constants.js:
```diff
@@ -422,3 +421,9 @@ export const APPIUM_CAPABILITES = [
+    method: 'GET',
+    host: 'localhost',
+    port: 4444,
+    path: '/status'
```
path should respect the `path` defined by user in config! For selenium v3 and some cloud providers it should be /wd/hub/status

I've just verified it locally

christian-bromann replied:
by default all drivers should start with path `/` if you change that we should assume you run `webdriver`. I modified the `getAutomationProtocol` method to now take this into consideration:

```js
/**
 * run WebDriver if hostname or port is set
 */
if (config.hostname || config.port || config.path || (config.user && config.key)) {
    return 'webdriver'
}
```

erwinheitzman commented on e2e/testrunner.test.js:
```diff
@@ -1,10 +1,28 @@
+        const rootPath = path.join(__dirname, 'wdio')
+        const logFiles = fs.readdirSync(rootPath).filter((file) => (
+            // only log files
+            file.endsWith('.log')
```
This filter does not work as `file.endsWith('.log')` needs to be returned here.

christian-bromann replied:
it returns because the surrounding `(` and `)`

christian-bromann replied:
made it more clear

christian-bromann commented on .travis.yml:
```diff
@@ -1,18 +1,93 @@
+cache:
+  directories:
+    - node_modules
+    - packages/devtools/node_modules
```
well we don't want to cache our build files

christian-bromann commented:
@mgrybyk @erwinheitzman addressed your comments

christian-bromann commented:
Moving forward as mentioned in the project committers channel.
